### PR TITLE
Now properly supports sealed traits & sealed enums

### PR DESCRIPTION
**Pull request by reactormonk**
Also properly smoke tested. Any way to generate the full json and run it against a linter?

### CODE REVIEW

**Comment by refried** [comment]
I got excited when I saw this but then I realized I don’t know what it is.  Is there any discussion or write-up or example?  Thanks.

**Comment by reactormonk** [comment]
Here's an example: https://gist.github.com/reactormonk/344c175c948777dc181b1074651aa7fd

**Comment by refried** [comment]
Ok I'm excited again.  This (better descriptions of custom data types) is something I wished for, and wanted to work on, but didn't have time / permission.

FYI I pasted the gist into http://editor.swagger.io/ 's online swagger editor/previewer and got three errors; one "not a valid status definition" and two "not a valid response definition".  I don't really know what they mean, and they may not have anything to do with this PR, but I thought I'd mention it.

**Comment by reactormonk** [comment]
Yeah, I fixed these. Updated the gist.

File: swagger/src/main/scala/org/http4s/rho/swagger/TypeBuilder.scala [hunk 1]
```diff
@@ -82,14 +82,17 @@ object TypeBuilder {
...
         case tpe@TypeRef(_, sym: Symbol, tpeArgs: List[Type]) if isSumType(sym) =>
           // TODO promote methods on sealed trait from children to model
           modelToSwagger(tpe, sfs).map(addDiscriminator(sym)).toSet.flatMap { model =>
-            val parent = RefModel(model.id + "Ref", model.id2 + "Ref", model.id2)
+            val refmodel = RefModel(model.id, model.id2, model.id2)
             val children =
-              sym.asClass.knownDirectSubclasses.flatMap( sub =>
-                  modelToSwagger(sub.asType.toType, sfs).map(composedModel(parent))
-                )
+              sym.asClass.knownDirectSubclasses.flatMap({ sub =>
```
Comment (bryce-anderson) [inline swagger/src/main/scala/org/http4s/rho/swagger/TypeBuilder.scala]: Remove one of the `(` or `{` layers from the `flatMap`.

File: swagger/src/main/scala/org/http4s/rho/swagger/TypeBuilder.scala [hunk 2]
```diff
@@ -82,14 +82,17 @@ object TypeBuilder {
 
           models ++ generics ++ children
 
+        case tpe@TypeRef(_, sym: Symbol, tpeArgs: List[Type]) if isObjectEnum(sym) =>
+          Set()
```
Comment (bryce-anderson) [inline swagger/src/main/scala/org/http4s/rho/swagger/TypeBuilder.scala]: `Set.empty` to be consistent.

File: swagger/src/main/scala/org/http4s/rho/swagger/models.scala [hunk 3]
```diff
@@ -578,6 +578,7 @@ object models {
     , access           : Option[String]   = None
     , vendorExtensions : Map[String, Any] = Map.empty
     , isArray          : Boolean          = false
+    , enums            : List[String]     = List()
```
Comment (bryce-anderson) [inline swagger/src/main/scala/org/http4s/rho/swagger/models.scala]: `List.empty`

**Comment by refried** [comment]
Just out of curiosity, since I haven’t dug into it, does this require Circe to work?

**Comment by reactormonk** [comment]
Nope. Just expects you to set the json codec accordingly.

**Comment by bryce-anderson** [comment]
These new features are looking pretty sweet. Great work @reactormonk.

**Pull request merged**